File: apps/server/src/api/health.ts
import process from 'node:process'
import { Elysia } from 'elysia'
import { isShuttingDown } from '@/state'

export const healthApi = new Elysia()
  .get('/health', ({ set }) => {
    if (isShuttingDown()) {
      set.status = 503
      return { status: 'shutting_down', version: process.env.VERSION ?? 'dev' }
    }
    return { status: 'ok', version: process.env.VERSION ?? 'dev' }
  }, {
    detail: { summary: 'Health check', tags: ['Health'] },
  })

File: internal/api/health.go
package api

import (
	"context"
	"net/http"
	"os"

	"github.com/danielgtaylor/huma/v2"
	"github.com/jkrumm/rollhook/internal/state"
)

type healthOutput struct {
	Status int
	Body   struct {
		Status  string `json:"status"`
		Version string `json:"version"`
	}
}

func RegisterHealth(api huma.API) {
	huma.Register(api, huma.Operation{
		OperationID: "get-health",
		Method:      http.MethodGet,
		Path:        "/health",
		Summary:     "Health check",
		Tags:        []string{"Health"},
	}, func(_ context.Context, _ *struct{}) (*healthOutput, error) {
		version := os.Getenv("VERSION")
		if version == "" {
			version = "dev"
		}
		out := &healthOutput{}
		out.Body.Version = version
		if state.IsShuttingDown() {
			out.Status = http.StatusServiceUnavailable
			out.Body.Status = "shutting_down"
		} else {
			out.Status = http.StatusOK
			out.Body.Status = "ok"
		}
		return out, nil
	})
}
